packages/@lwc/style-compiler/src/serialize.ts

```diff
@@ -174,34 +175,35 @@ function serializeCss(result: Result, collectVarFunctions: boolean): string {
     return generateExpressionFromTokens(tokens);
 }
 
-// TODO [#1288]: this code needs refactor, it could be simpler by using a native post-css walker
 function tokenizeCssSelector(data: string): Token[] {
     data = data.replace(/( {2,})/gm, ' '); // remove when there are more than two spaces
-    const tokens: Token[] = [];
-    let pos = 0;
-    let next = 0;
-    const max = data.length;
-
-    while (pos < max) {
-        if (data.indexOf(`[${HOST_ATTRIBUTE}]`, pos) === pos) {
-            tokens.push({ type: TokenType.identifier, value: HOST_SELECTOR_IDENTIFIER });
 
-            next += HOST_ATTRIBUTE.length + 2;
-        } else if (data.indexOf(`[${SHADOW_ATTRIBUTE}]`, pos) === pos) {
-            tokens.push({ type: TokenType.identifier, value: SHADOW_SELECTOR_IDENTIFIER });
+    const tokens: Token[] = [];
 
-            next += SHADOW_ATTRIBUTE.length + 2;
+    // Directly use the SelectorParser from postcss-selector-parser because this is the only
+    // way to get the tokens without the full AST. We want the tokens rather than the AST
+    // because we're outputting a flat array, not a tree.
+    const parser = new SelectorParser(data);
+
+    for (let i = 0; i < parser.tokens.length; i++) {
+        const token = parser.tokens[i];
+        // token[5] is the start index, token[6] is the end character
+        // See: https://github.com/postcss/postcss-selector-parser/blob/e64151e/src/tokenize.js#L280-L288
```

Comment:
I don't know much about CSS or the CSS parser, but I agree with @ekashida : This PR relies on implementation details from postcss-selector-parser; which I do not like.

@nolanlawson, what's the advantage of using this approach instead of our own implementation? how complicated is it to use the public API of postcss-selector-parser?

Reply:
I looked into using `postcss-selector-parser`'s public API, but it's too ill-suited to what we're trying to do, since it's an AST and what we want is a flat array of tokens. We would basically have to reverse-engineer the tokens from the AST. Sadly `postcss-selector-parser` also doesn't seem to expose its tokenizer anywhere.

I'm fine with either decision we choose. It's probably faster to use our own string parser. But if we decide not to change anything, we should remove the TODO.